atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java

```diff
@@ -0,0 +1,132 @@
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.stream;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+import com.google.common.base.Preconditions;
+
+public final class BlockConsumingInputStream extends InputStream {
+    private final BlockGetter blockGetter;
+    private final long numBlocks;
+    private final int blocksInMemory;
+
+    private long nextBlockToRead;
+
+    private byte[] buffer;
+    private int positionInBuffer;
+
+    public static BlockConsumingInputStream create(
+            BlockGetter blockGetter,
+            long numBlocks,
+            int blocksInMemory) throws IOException {
+        ensureExpectedArraySizeDoesNotOverflow(blockGetter, blocksInMemory);
+        BlockConsumingInputStream stream = new BlockConsumingInputStream(blockGetter, numBlocks, blocksInMemory);
+        stream.refillBuffer();
+        return stream;
+    }
+
+    private static void ensureExpectedArraySizeDoesNotOverflow(BlockGetter blockGetter, int blocksInMemory) {
+        int expectedLength = blockGetter.expectedLength();
+        Preconditions.checkArgument(blocksInMemory < Integer.MAX_VALUE / expectedLength,
+                "Promised to load too many blocks into memory. The underlying buffer is stored as a byte array, "
+                        + "so can only fit Integer.MAX_VALUE bytes. The supplied BlockGetter expected to produce "
+                        + "blocks of %s bytes, so %s of them would cause the array to overflow.",
+                blockGetter.expectedLength(),
+                blocksInMemory);
+    }
+
+    private BlockConsumingInputStream(BlockGetter blockGetter, long numBlocks, int blocksInMemory) {
+        this.blockGetter = blockGetter;
+        this.numBlocks = numBlocks;
+        this.blocksInMemory = blocksInMemory;
+        this.nextBlockToRead = 0L;
+        this.positionInBuffer = 0;
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (positionInBuffer < buffer.length) {
+            return buffer[positionInBuffer++] & 0xff;
+        }
+
+        if (nextBlockToRead < numBlocks) {
+            boolean reloaded = refillBuffer();
+            if (!reloaded) {
+                return -1;
+            }
+
+            return buffer[positionInBuffer++] & 0xff;
+        }
+
+        return -1;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        Preconditions.checkNotNull(b, "Cannot read into a null array!");
+        if (off < 0 || len < 0 || len > b.length - off) {
```

Comment:
That check comes from the javadoc of InputStream, which I'd rather not contradict.